### PR TITLE
feat(cli): validate snapshots

### PR DESCRIPTION
**Pull request by y-lakhdar**
## Proposed changes

Validate a snapshot right after it is being pushed to an organization.
This PR does not handle all the invalid cases. It considers the snapshot valid if we get a `SUCESS` result code from the API. Otherwise, the snapshot is considered invalid. [CDX-362](https://coveord.atlassian.net/browse/CDX-362) will handle all the possible invalid states of a snapshot.

There is also a patch for the platform-client until the pull request [322](https://github.com/coveo/platform-client/pull/322) is merged. Without that change, it is not possible to get the status report of a snapshot.

## Testing
I split the snapshot tests into 2 files:
1. snapshot.spec.ts
2. snapshotFactory.spec.ts

- [x] Unit Tests:
- [x] Manual Tests: Manually tested valid and invalid snapshots.

## Acceptance Criteria 
- [x] Merge https://github.com/coveo/platform-client/pull/322

-----
CDX-358

### CODE REVIEW

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 1]
```diff
@@ -8,7 +15,19 @@ export class Snapshot {
...
   async validate() {
-    // TODO: CDX-358: Validate snapshot
+    await this.client.dryRun(this.snapshotId, {
+      deleteMissingResources: false, // TODO: CDX-361: Add flag to support missing resources deletion
+    });
+
+    await this.waitUntilIdle();
+  }
+
+  isValid() {
```
Comment (olamothe) [inline packages/cli/src/lib/snapshot/snapshot.ts]: The class interface here is a bit weird, because you have to call `validate()`, then you can call `isValid()` afterwards.

As a user of this class, you could very well try to call `isValid()` without first doing `validate()`, and then you'd get incorrect results.

I think we could refactor this so `validate()` returns true or false, and we could remove the `isValid()` method.

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 2]
```diff
@@ -28,4 +47,53 @@ export class Snapshot {
   async delete() {
     // TODO: CDX-359: Delete snapshot once previewed
   }
+
+  get lastestReport(): ResourceSnapshotsReportModel {
```
Comment (olamothe) [inline packages/cli/src/lib/snapshot/snapshot.ts]: By refactoring it so `validate()` control the whole flow, instead of 2 method, this check would go away.

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 3]
```diff
@@ -28,4 +47,53 @@ export class Snapshot {
...
+    }
+    return this.model.reports.slice(-1)[0];
+  }
+
+  private get snapshotId() {
+    return this.model.id;
+  }
+
+  private async refreshSnapshotData() {
+    this.model = await (this.client.get as any)(this.model.id, {
```
Comment (olamothe) [inline packages/cli/src/lib/snapshot/snapshot.ts:65]: For the `as any` cast, is this a bug in platform-client ? Should this be a TODO ?

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 4]
```diff
@@ -28,4 +47,53 @@ export class Snapshot {
...
+      includeReports: false,
+    });
+  }
+
+  private async isIdle() {
+    await this.refreshSnapshotData();
+    const lastReport = this.model.reports?.slice(-1)[0];
```
Comment (olamothe) [inline packages/cli/src/lib/snapshot/snapshot.ts]: That seems very similar to `get lastestReport()` ? 🤔

File: packages/cli/src/lib/snapshot/snapshot.ts (hunk 3 above)
Comment (y-lakhdar) [reply packages/cli/src/lib/snapshot/snapshot.ts]: yes, a  bug in the platform-client. I created a Pull request and added the link to the the jira in the comments

File: packages/cli/src/lib/snapshot/snapshot.ts (hunk 4 above)
Comment (y-lakhdar) [reply packages/cli/src/lib/snapshot/snapshot.ts]: yeah, that was a bad commit. I replaced it with the get latestReport :P

File: packages/cli/src/lib/snapshot/snapshot.ts (hunk 3 above)
Comment (y-lakhdar) [reply packages/cli/src/lib/snapshot/snapshot.ts]: Pull request : https://github.com/coveo/platform-client/pull/322

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 5]
```diff
@@ -24,7 +55,52 @@ export class Snapshot {
     // TODO: CDX-347 Display Expanded preview
   }
 
-  async delete() {
-    // TODO: CDX-359: Delete snapshot once previewed
+  private get lastestReport(): ResourceSnapshotsReportModel {
```
Comment (olamothe) [inline packages/cli/src/lib/snapshot/snapshot.ts:58]:
```suggestion
  private get latestReport(): ResourceSnapshotsReportModel {
```

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 6]
```diff
@@ -24,7 +55,52 @@ export class Snapshot {
...
+    if (this.model.reports === undefined) {
+      throw new Error(
+        `No detailed report found for the snapshot ${this.snapshotId}`
+      );
+    }
+    return this.model.reports.slice(-1)[0];
```
Comment (olamothe) [inline packages/cli/src/lib/snapshot/snapshot.ts:64]: Anything we might want to handle as far as reports array being empty ?
Is it even possible ?

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 7]
```diff
@@ -24,7 +55,52 @@ export class Snapshot {
...
+
+  private async refreshSnapshotData() {
+    this.model = await this.snapshotClient.get(this.model.id, {
+      includeReports: true,
+    });
+  }
+
+  private async isIdle() {
+    await this.refreshSnapshotData();
+    return new Promise<void>((resolve) => {
```
Comment (olamothe) [inline packages/cli/src/lib/snapshot/snapshot.ts:83]: I'm not sure I understand why this is wrapped in a Promise ?

There's no rejection that would trigger the `waitUntilIdle` backoff mechanism, so... 🤔

**Review by olamothe — APPROVED** [review]

**Comment by olamothe** [comment]
💯  Gj !

File: packages/cli/src/lib/snapshot/snapshot.ts (hunk 7 above)
Comment (y-lakhdar) [reply packages/cli/src/lib/snapshot/snapshot.ts:83]: My bad. I forgot the `reject()`. I am not sure that is a good usage of the `exponential-backoff` package (as opposed to async-retry).

File: packages/cli/src/lib/snapshot/snapshot.ts (hunk 6 above)
Comment (y-lakhdar) [reply packages/cli/src/lib/snapshot/snapshot.ts:64]: Probably a question to the backend team.
I included the `includeReport` query parameter on every GET request to the snapshot API. 

I'll ask for confirmation

File: packages/cli/patches/@coveord+platform-client+18.15.2.patch [hunk 8]
```diff
@@ -0,0 +1,54 @@
+diff --git a/node_modules/@coveord/platform-client/dist/definitions/resources/ResourceSnapshots/ResourceSnapshots.d.ts b/node_modules/@coveord/platform-client/dist/definitions/resources/ResourceSnapshots/ResourceSnapshots.d.ts
```
Comment (louis-bompart) [inline packages/cli/patches/@coveord+platform-client+18.15.2.patch:1]: I'm really not in favor of doing a patch on dependencies where we have a bit of control (I mean, it's in the cie)
I can't approve this PR with this.

File: packages/cli/src/lib/snapshot/snapshot.spec.ts [hunk 9]
```diff
@@ -1,28 +1,75 @@
...
+    id: snapshotId,
+    createdBy: 'user@coveo.com',
+    createdDate: 1622555047116,
+    targetId: orgId,
+    developerNote: 'hello',
+    reports: reports,
+    synchronizationReports: [],
+    contentSummary: {EXTENSION: 1, FIELD: 11},
+  };
 };
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.spec.ts:73]: Not necessarly prettier compliant (i.e. might need a alt-shift-f in vscode)
But yeah, useless to do a full block function imo.

```suggestion
): ResourceSnapshotsModel => {
  {
    id: snapshotId,
    createdBy: 'user@coveo.com',
    createdDate: 1622555047116,
    targetId: orgId,
    developerNote: 'hello',
    reports: reports,
    synchronizationReports: [],
    contentSummary: {EXTENSION: 1, FIELD: 11},
  };
```

File: packages/cli/src/lib/snapshot/snapshot.spec.ts [hunk 10]
```diff
@@ -46,49 +95,101 @@ const doMockAuthenticatedClient = () => {
...
+        targetOrgId,
+        snapshotId,
+        [successReport]
+      );
+
+      mockedGetSnapshot.mockResolvedValueOnce(futureSnapshotState);
+    };
+
+    beforeEach(async () => {
+      doMockGetSnapshotWithoutError();
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.spec.ts:178]: await?
```suggestion
      await doMockGetSnapshotWithoutError();
```

File: packages/cli/src/lib/snapshot/snapshot.spec.ts [hunk 11]
```diff
@@ -46,49 +95,101 @@ const doMockAuthenticatedClient = () => {
...
+        [errorReport]
+      );
+
+      mockedGetSnapshot.mockResolvedValueOnce(futureSnapshotState);
+    };
 
     beforeEach(async () => {
-      await SnapshotFactory.createFromZip(pathToZip, 'my-target-org');
-    });
+      doMockGetSnapshotInError();
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.spec.ts:119]: await?
```suggestion
      await doMockGetSnapshotInError();
```

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 12]
```diff
@@ -24,7 +55,54 @@ export class Snapshot {
...
+          ResourceSnapshotsReportStatus.Aborted,
+          ResourceSnapshotsReportStatus.Completed,
+        ].includes(this.latestReport.status)
+      ) {
+        resolve();
+      } else {
+        reject();
+      }
+    });
+  }
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.ts:95]: Mhhh. Not sure here, what `isIdle` for you in layman terms?

**Review by louis-bompart — CHANGES_REQUESTED** [review]
Some pet peeves, but mostly blocking approval due to the patch (tho it's fine for ongoing dev)

File: packages/cli/patches/@coveord+platform-client+18.15.2.patch (hunk 8 above)
Comment (y-lakhdar) [reply packages/cli/patches/@coveord+platform-client+18.15.2.patch:1]: Blocked by https://github.com/coveo/cli/pull/240/

File: packages/cli/src/lib/snapshot/snapshot.ts (hunk 12 above)
Comment (y-lakhdar) [reply packages/cli/src/lib/snapshot/snapshot.ts:95]: Meaning that snapshot is not in a pending state.
In the platform, that pending state is translated with a loading animation (see image).
image

The role of that `isIdle` function is to keep fetching the snapshot status until the report indicates a failure or a success.
But I agree, we can change that name :P

File: packages/cli/src/lib/snapshot/snapshot.spec.ts (hunk 11 above)
Comment (y-lakhdar) [reply packages/cli/src/lib/snapshot/snapshot.spec.ts:118]: mmh non. I think the doMockGetSnapshotInError should not return a promise

Comment (y-lakhdar) [reply packages/cli/src/lib/snapshot/snapshot.spec.ts:117]: Good catch tho

File: packages/cli/src/lib/snapshot/snapshot.ts (hunk 12 above)
Comment (louis-bompart) [reply packages/cli/src/lib/snapshot/snapshot.ts:95]: > Meaning that the snapshot is not being processed by some kind of operation (validation, deployment, being applied). That it had reached a finish state.

Is there a defined status for that in `ResourceSnapshotsReportStatus` ?

The thing that I found odd tbh is that `isIdle` resolves if the status is Aborted or Completed.

I interpret that as `isIdle` is 'true' if the status is Completed or Aborted. Which is strange imho

also resolve, reject, meh. I think a resolve w/ a boolean would be more semantically correct with the `isXX` name.

Comment (y-lakhdar) [reply packages/cli/src/lib/snapshot/snapshot.ts:95]: `isIdle` doesn't mean the snapshot is valid. It only means that it has settled from its last operation. We ca also go the other way around and have a method `isPending` that will check if the snapshot has one of the following values `Pending`, `InProgress`. 

I agree about the resolve reject. The other way would be to throw an error if the snapshot is still not idle

Comment (louis-bompart) [reply packages/cli/src/lib/snapshot/snapshot.ts:95]: > `isIdle` doesn't mean the snapshot is valid. It only means that it has settled from its last operation. We can also go the other way around and have a method `isPending` that will check if the snapshot is in one of the following states `PENDING`, `IN_PROGRESS`.
> 
> I agree about the resolve reject. The other way would be to throw an error if the snapshot is still not idle

Either way, I would name the method maybe `isComplete` or `isDone`
Idle I interpret this as "there's something that will go on, but not right now"

Comment (y-lakhdar) [reply packages/cli/src/lib/snapshot/snapshot.ts:95]: `isDone` it is.

File: packages/cli/src/lib/snapshot/snapshot.spec.ts [hunk 13]
```diff
@@ -1,92 +1,190 @@
...
+    id: snapshotId,
+    updatedDate: 1622555847000,
+    type: type,
+    status: status,
+    resultCode: resultCode,
+    resourcesProcessed: 12,
+    resourceOperations: {},
+    resourceOperationResults: {},
+  };
+};
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.spec.ts:38]:
```suggestion
): ResourceSnapshotsReportModel => {
  id: snapshotId,
  updatedDate: 1622555847000,
  type: type,
  status: status,
  resultCode: resultCode,
  resourcesProcessed: 12,
  resourceOperations: {},
  resourceOperationResults: {},
};
```

File: packages/cli/src/lib/snapshot/snapshot.spec.ts [hunk 14]
```diff
@@ -1,92 +1,190 @@
...
+
+const getSuccessReport = (snapshotId: string): ResourceSnapshotsReportModel => {
+  return getReport(
+    snapshotId,
+    ResourceSnapshotsReportType.DryRun,
+    ResourceSnapshotsReportStatus.Completed,
+    ResourceSnapshotsReportResultCode.Success
+  );
+};
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.spec.ts:47]:
```suggestion
const getSuccessReport = (snapshotId: string): ResourceSnapshotsReportModel =>
  getReport(
    snapshotId,
    ResourceSnapshotsReportType.DryRun,
    ResourceSnapshotsReportStatus.Completed,
    ResourceSnapshotsReportResultCode.Success
  );
```

File: packages/cli/src/lib/snapshot/snapshot.spec.ts [hunk 15]
```diff
@@ -1,92 +1,190 @@
...
+
+const getErrorReport = (snapshotId: string): ResourceSnapshotsReportModel => {
+  return getReport(
+    snapshotId,
+    ResourceSnapshotsReportType.DryRun,
+    ResourceSnapshotsReportStatus.Completed,
+    ResourceSnapshotsReportResultCode.ResourcesInError
+  );
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.spec.ts:55]:
```suggestion
const getErrorReport = (snapshotId: string): ResourceSnapshotsReportModel =>
  getReport(
    snapshotId,
    ResourceSnapshotsReportType.DryRun,
    ResourceSnapshotsReportStatus.Completed,
    ResourceSnapshotsReportResultCode.ResourcesInError
  );
```

**Review by louis-bompart — COMMENTED** [review]
Small suggestions, it's flavouring tho, so feel free to ignore them!
Other than that, as mentioned earlier yeah there's the `isIdle`/`isDone` thing.
Rest is A-OK, good job! 👏

File: packages/cli/patches/@coveord+platform-client+18.15.2.patch (hunk 8 above)
Comment (y-lakhdar) [reply packages/cli/patches/@coveord+platform-client+18.15.2.patch]: Not blocked anymore

**Review by louis-bompart — APPROVED** [review]
